# Курс "Практикум по математической статистике"
# 3 курс ФПМИ МФТИ, осень 2022
## Домашнее задание 2. Методы нахождения оценок

Мы предлагаем выполнять задания прямо в этом ноутбуке. Пожалуйста, не стирайте условия задач.

Настоятельно рекомендуемая форма оформления домашних заданий — это Jupyter Notebook c:

* условием задачи,
* решением (если требуется некоторый теоретический вывод),
* описанием плана решения, который потом реализуется в коде, 
* собственно кодом, 
* построенными графиками (если это требуется) и **выводом**, который как правило должен заключаться в объяснении практических результатов с использованием теоретических фактов. ***Вывод требуется даже в том случае, если в условии об этом явно не сказано!***
* некоторыми другими вещами, если об этом будет указано в задании.

Оценка за каждую задачу складывается из правильного выполнения всех этих пунктов. Закрывая на них глаза, вы сознательно понижаете свою оценку.

Каждая задача оценивается в 10 баллов, если не оговорено иного.

Загрузим все необходимые датасеты. Если что-то пошло не так, то просто скачайте файлы по ссылке вручную.

In [59]:
!pip install -q gdown
!gdown https://drive.google.com/uc?id=1fMQ0H-_E4U25XHB2SH7ryoZPLG2MHlLQ
!gdown https://drive.google.com/uc?id=1cJywRii7wBZa0B2uAvvu56JFCLPnlOSs

Downloading...
From: https://drive.google.com/uc?id=1fMQ0H-_E4U25XHB2SH7ryoZPLG2MHlLQ
To: /home/ashot/Documents/NOV/Учеба/mipt/Питон/HW/hw2/3.Основные_методы_нахождения_оценок/Cauchy.csv
100%|███████████████████████████████████████| 18.7k/18.7k [00:00<00:00, 483kB/s]
Downloading...
From: https://drive.google.com/uc?id=1cJywRii7wBZa0B2uAvvu56JFCLPnlOSs
To: /home/ashot/Documents/NOV/Учеба/mipt/Питон/HW/hw2/3.Основные_методы_нахождения_оценок/Weibull.csv
100%|██████████████████████████████████████| 17.9k/17.9k [00:00<00:00, 18.9MB/s]


In [60]:
import pandas as pd
import numpy as np
from scipy import stats as sps

from matplotlib import pyplot as plt
import seaborn as sns

sns.set(style="darkgrid", font_scale=1.4)

## Задача 1

На высоте 1 метр от поверхности Земли закреплено устройство,
которое периодически излучает лучи на поверхность Земли (считайте,
что поверхность Земли представляет из себя прямую). Пусть $l$ —
перпендикуляр к поверхности Земли, опущенный из точки, в которой
закреплено устройство. Угол к прямой $l$ (под которым происходит
излучение) устройство выбирает случайно из равномерного
распределения на от\-резке $\left(-\frac{\pi}{2}, \frac{\pi}{2}\right)$ (все выборы
осуществляются независимо). В этих предположениях точки пересечения
с поверхностью имеют распределение Коши с плотностью $p(x) =
\frac{1}{\pi(1 + (x-x_0)^2)}.$ Неизвестный параметр сдвига $x_0$
соответствует проекции точки расположения устройства на поверхность
Земли (направление оси и начало координат на поверхности Земли
выбраны заранее некоторым образом независимо от расположения
устройства). В файле Cauchy.csv находятся координаты точек
пересечения лучей с поверхностью Земли.

 Оцените параметр сдвига
методом максимального правдоподобия 
* по первым 10 измерениям
* по первым 100 измерениям
* по всей выборке. 

Оценку произведите по сетке (т.е. возьмите набор
точек с некоторым шагом и верните ту, на которой достигается
максимум функции правдоподобия). Известно, что параметр сдвига принадлежит интервалу $[-1000, 1000].$ Выберите шаг равным 0.01. Интервал можете итеративно уменьшать, но не стоит делать его длину меньше 50.


In [61]:
# Записываем выборку из файла в список sample
f = open('Cauchy.csv.xls', 'r')
text = f.read()
lines = text.split('\n')
sample = np.array([float(lines[n]) for n in range(len(lines) - 1)])

In [62]:
# Логарифм функции правдоподобия (сумма логарифмов значений плотности на каждом значении из sample).
def likehood(sample, loc_):
    return np.sum(cauchy.logpdf(sample, loc=loc_, scale=1))

In [63]:
# Ищем максимум логарифмической функции правдоподобия по loc_, перебирая его от -1000 до 1000 с шагом 0.01.
# Выводится значение параметра сдвига, на котором достигается максимум.

# ценка по всей выборке
from scipy.stats import cauchy
left_bound = -1000
right_bound = 1000
step = 0.01
estimation = left_bound + step * (np.argmax(np.array([likehood(sample, loc_) 
                                               for loc_ in np.arange(left_bound, right_bound, step)])))

In [64]:
print(estimation)

-208.03999999999996


In [65]:
# оценка по первым 10 измерениям
sample = np.array(sample[:10])

left_bound = -1000
right_bound = 1000
step = 0.01
estimation = left_bound + step * (np.argmax(np.array([likehood(sample, loc_) 
                                               for loc_ in np.arange(left_bound, right_bound, step)])))
print(estimation)


-208.14


In [66]:
#оценка по первым 100 измерениям
sample = np.array(sample[:100])

left_bound = -1000
right_bound = 1000
step = 0.01
estimation = left_bound + step * (np.argmax(np.array([likehood(sample, loc_) 
                                               for loc_ in np.arange(left_bound, right_bound, step)])))
print(estimation)

-208.14


### Вывод

Мы оценили параметр сдвига в распределении Коши методом правдоподобия по всей выборке, по первым 10 измерениям и по первым 100 измерениям. В первом случае оценка параметра 𝑥0=−208.04, во втором — 𝑥0=-208.14, в третьем — 𝑥0=-208.14. Как мы видим, результаты хоть и различаются, но не намного.

Сравните полученные результаты с `sps.cauchy.fit`

## Задача 2
В банкомате "Тинькофф" в Новом Корпусе МФТИ каждую минуту подсчитывается баланс по
сравнению с началом дня (6 часов утра). В полночь работники банка
измеряют две величины: $X^1$ – максимальное значение баланса за
день, $X^2$ – значение баланса в полночь. Считается, что величина
$X = X^1 - X^2$ имеет распределение Вейбулла с функцией распределения $F(x) = 1 - e^{-x^\gamma} (x > 0),$ где $\gamma > 0$ –
параметр формы. В течение 10 лет каждый день банк проводил
измерение величины $X,$ получив в результате выборку $X_1, \ldots,
X_{3652}.$ В файле Weibull.csv находятся соответствующие измерения.

In [67]:
sample = np.loadtxt("Weibull.csv")

Проведем небольшой предварительный анализ. Итак, если наши данные распределены согласно распределению Вейбулла, то справедливы следующие рассуждения:
$$
\begin{aligned} F(x) &=1-e^{-(x)^{\gamma}} \\-\ln (1-F(x)) &=x^{\gamma} \\ \underbrace{\ln (-\ln (1-F(x)))}_{y^{\prime}} &=\underbrace{\gamma \ln x}_{\text {kx }^{\prime}} \end{aligned}
$$
А значит и
$$
\underbrace{\ln (-\ln (1-\hat{F}(x)))}_{y^{\prime}} ≈\underbrace{\gamma \ln x}_{\text {kx }^{\prime}}
$$
Подсчитайте эмпирическую функцию распределения и 
$$y\prime = \ln (-\ln (1-\hat{F}(x))) \quad x' = ln x$$
 где $x$ – элементы исходной выборки. Постройте график (`plt.scatter`) выделив данные за первые четыре года красным цветом (`sample_2[:1461]`), остальные синим (`sample_2[1461:]`). Не забудьте про `alpha=0.05` и легенду. Такой график называется `Weibull plot` и является аналогом `qqplot` для распределения Вейбулла. 


In [68]:
from statsmodels.distributions.empirical_distribution import ECDF
from scipy.stats import weibull_min

# Логарифм функции правдоподобия (сумма логарифмов значений плотности на каждом значении из sample).
# log_shape - десятичный логарифм параметра формы. Аргументтом logpdf является число 10 в степени log_shape.
def likehood(sample, log_shape):
    return np.sum(weibull_min.logpdf(sample, 10 ** log_shape))

Сделайте вывод.

**Вывод**

Оцените параметр формы методом максимального правдоподобия 
* по
первым 4 годам; 
* по всей выборке.
Оценку произведите по сетке (в
логарифмической шкале). Известно, что $\log_{10}\gamma \in [-2, 2].$
Выберите шаг равным $10^{-3}.$

In [69]:
# оценка по всей выборке

# Ищем максимум логарифмической функции правдоподобия по shape в логарифмическом масштабе, перебирая его от -2 до 2 с шагом 0.001.
# Выводится значение параметра формы, на котором достигается максимум.
left_bound = -2
right_bound = 2
step = 0.001
estimation = 10 ** (left_bound + step * (np.argmax(np.array([likehood(sample, log_shape) 
                                               for log_shape in np.arange(left_bound, right_bound, step)]))))
print(estimation)

0.01


In [70]:
# оценка по первым четырем годам

# Изменяем выборку.
sample = sample[:(365 * 4)]

left_bound = -2
right_bound = 2
step = 0.001
estimation = 10 ** (left_bound + step * (np.argmax(np.array([likehood(sample, log_shape) 
                                               for log_shape in np.arange(left_bound, right_bound, step)]))))
print(estimation)

0.01


Сравните результаты с `sps.weibull.fit(sample_2, fscale=1, floc=0)`

Постройте график $y = \gamma \cdot x$ для всех полученных $\gamma$ (`plt.plot`) и `scatter plot` из предыдущего пункта ($y\prime \ x\prime$). Хорошо ли линии соответствуют выборке? Как вы думаете, почему?


# Вывод

Мы оценили параметр формы в распределении Вейбулла методом правдоподобия по всей выборке и по ее первым 1460 значениям. В обоих случаях оценка параметра сдвига оказалась равной $\gamma^* = 0.01$.

## Задача 3

Сгенерируйте выборки $X_1, \ldots, X_N$ из $\mathit{N}(0, \theta)$, $U(0, \theta)$, $\Gamma(1, \theta)$ (**параметризация $k, \theta$**), $\theta = 3$ для всех распределений $(N = 1000).$ Для всех $n \leq N$ посчитайте значения оценок (по выборке $X_1, \ldots X_n$) методом моментов. Постройте график ошибки оценки от реального значения ($|\hat\theta - \theta|_{l_1}$) относительно размера выборки.

In [72]:
# YOUR_CODE_GOES_HERE

### Бутстреп

Для реальных данных часто сложно подобрать распределение и нужную параметризацию относительно $\theta$. Кроме того на практике сложно посчитать дисперсию оценки (для этого хотя бы нужно знать распределение, из которого пришла выборка). На помощь в таких случаях приходит **бутстреп**.

Идея очень простая. Давайте возьмем нашу выборку размера $N$ и сгенерируем из нее еще $K$ выборок. Более формально для каждой бутстрепной выборки N раз будем выбирать элементы из исходной выборки с возвращением. Полученная таким образом выборка будет содержать $\approx 63$% уникальных элементов, но это не страшно. Для всех $K$ выборок посчитаем оценку $\hat\theta$. Таким образом мы получим $K$ оценок параметра. Можно показать, что если размер бутстрепных выборок и исходной совпадают, то оценка дисперсии $s^2(\hat\theta)$, полученная из $K$ оценок, будет *хорошей*.

Для каждого распределения из предыдущего пункта (Пожалуйста, не пишите цикл по распределениям. Сделайте три отдельные ячейки) для каждого $K$ из `[10] + [50] + list(range(100, 1001, 100))` сгенерируйте K бутстрепных выборок и посчитайте дисперсию бутстрепных оценок и посчитайте среднее по K выборкам. Размер бутстрепной выборки сделайте равным $K$, незабудьте уменьшить размер исходной выборки до $K$. Постройте график следующим образом: по оси $x$ отложите значения $K$, красной линией обозначьте среднее значение $\hat\theta$ бутстрепных выборок для каждого $K$. Посчитайте стандартное отклонение оценки для каждого $K$ и закрасьте интервал $\mu(k) \pm \sigma(\hat\theta)$ (`plt.fillbetween`).



In [73]:
# YOUR_CODE_GOES_HERE